pointcloud/basic_octree_test.go

```diff
@@ -598,3 +598,87 @@ func testPCDToBasicOctree(t *testing.T, artifactPath string) {
 
 	validateBasicOctree(t, basicOct, basicOct.center, basicOct.sideLength)
 }
+
+func TestCachedMaxProbability(t *testing.T) {
+	center := r3.Vector{X: 0, Y: 0, Z: 0}
+	side := 2.0
+
+	octree, err := createNewOctree(center, side)
+	test.That(t, err, test.ShouldBeNil)
+
+	pointsAndData := []PointAndData{
+		{P: r3.Vector{X: 0, Y: 0, Z: 0}, D: NewValueData(2)},
+		{P: r3.Vector{X: .5, Y: 0, Z: 0}, D: NewValueData(3)},
+		{P: r3.Vector{X: .5, Y: 0, Z: .5}, D: NewValueData(10)},
+		{P: r3.Vector{X: .5, Y: .5, Z: 0}, D: NewValueData(1)},
+		{P: r3.Vector{X: .55, Y: .55, Z: 0}, D: NewValueData(4)},
+		{P: r3.Vector{X: -.55, Y: -.55, Z: 0}, D: NewValueData(5)},
+		{P: r3.Vector{X: .755, Y: .755, Z: 0}, D: NewValueData(6)},
+	}
+
+	err = addPoints(octree, pointsAndData)
+	test.That(t, err, test.ShouldBeNil)
+
+	validateBasicOctree(t, octree, octree.center, octree.sideLength)
+
+	// test that we can query the tree for max prob
+	mp, err := getMaxProb(octree, r3.Vector{X: 0, Y: 0, Z: 0})
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, mp, test.ShouldEqual, 0.02)
+
+	mp, err = getMaxProb(octree, r3.Vector{X: .755, Y: .755, Z: 0})
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, mp, test.ShouldEqual, 0.06)
+
+	_, err = getMaxProb(octree, r3.Vector{X: -.755, Y: -.755, Z: 0})
+
+	test.That(t, err.Error(), test.ShouldEqual, "point not found in passed in octree")
+}
+
+// validateProbability is a helper function that recursively checks that max probabilities are set as expected.
+func validateProbability(t *testing.T, octree *BasicOctree, leafMPCMap map[r3.Vector]float64) (float64, error) {
+	t.Helper()
+	misMatch := errors.New("node's MPC does not match the expected value")
+	switch octree.node.nodeType {
+	case internalNode:
+		allChildrenProbs := []float64{}
+		// recur on children
+		for _, child := range octree.node.children {
+			p, err := validateProbability(t, child, leafMPCMap)
+			if err != nil {
+				return 0, err
+			}
+			allChildrenProbs = append(allChildrenProbs, p)
+		}
+		computedMPC := getLargestNum(allChildrenProbs)
+		actualMPC := octree.node.maxProb
+		if computedMPC != actualMPC {
+			return 0, misMatch
+		}
+		return actualMPC, nil
+
+	case leafNodeFilled:
+		p := octree.node.point.P
```

Comment:
[minor] this check should just be that d/100 == maxProb, the current implementation is doing this but with the extra step of storing the variable and passing it around